processing/src/main/java/io/druid/query/monomorphicprocessing/SpecializationService.java

```diff
@@ -0,0 +1,388 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query.monomorphicprocessing;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.io.ByteStreams;
+import io.druid.concurrent.Execs;
+import io.druid.java.util.common.logger.Logger;
+import org.objectweb.asm.ClassReader;
+import org.objectweb.asm.ClassVisitor;
+import org.objectweb.asm.ClassWriter;
+import org.objectweb.asm.commons.ClassRemapper;
+import org.objectweb.asm.commons.SimpleRemapper;
+import sun.misc.Unsafe;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.reflect.Field;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Manages class specialization during query processing.
+ * Usage:
+ *
+ * String runtimeShape = stringRuntimeShape.of(bufferAggregator);
+ * SpecializationState<ProcessingAlgorithm> specializationState = SpecializationService.getSpecializationState(
+ *   ProcessingAlgorithmImpl.class,
+ *   runtimeShape
+ * );
+ * ProcessingAlgorithm algorithm = specializationState.getSpecializedOrDefault(new ProcessingAlgorithmImpl());
+ * long loopIterations = new ProcessingAlgorithmImpl().run(bufferAggregator, ...);
+ * specializationState.accountLoopIterations(loopIterations);
+ *
+ * ProcessingAlgorithmImpl.class, passed as prototypeClass to {@link #getSpecializationState} methods must have public
+ * no-arg constructor and must be stateless (no fields).
+ *
+ * @see SpecializationState
+ */
+public final class SpecializationService
+{
+  private static final Logger LOG = new Logger(SpecializationService.class);
+
+  private static final Unsafe UNSAFE;
+  static {
+    try {
+      Field theUnsafe = Unsafe.class.getDeclaredField("theUnsafe");
+      theUnsafe.setAccessible(true);
+      UNSAFE = (Unsafe) theUnsafe.get(null);
+    } catch (Exception e) {
```

Comment:
I mean, a catch block should start on a new line. Please find ```<option name="CATCH_ON_NEW_LINE" value="true" />``` at [here](https://raw.githubusercontent.com/druid-io/druid/master/druid_intellij_formatting.xml).